File: Linked_Lists/141_linked_list_cycle/141.c
#include "../list.h"
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>



bool Traverse(struct SingleListNode* slow, struct SingleListNode* fast) {

    if (slow == NULL || slow -> next == NULL) {
        return false;
    }

    if (fast == NULL || fast -> next == NULL) {
        return false;
    }

    if (slow == fast) {
        return true;
    }
    return Traverse(slow -> next, fast -> next -> next);
}


bool hasCycle(struct SingleListNode* head) {
    if (head == NULL) {
        return false;
    }
    return Traverse(head, head -> next);
}


int main() {

    int array[] = {3, 2, 0, -4};
    int arraySize = 4;
    struct SingleListNode* head = build_list(array, arraySize);
    printf("list : ");
    print_nodes(head);

    head -> next -> next -> next = head;

    printf("has cycle : %d\n", hasCycle(head));
    return 0;
}

File: Linked_Lists/141_linked_list_cycle/141.py
import sys
sys.path.append(r'/home/mount/Linked_Lists/')
import list




def hasCycle(head):
    
    if head is None:
        return False
    
    def Traverse(slow, fast):

        if slow == None or slow.next == None:
            return False
        if fast == None or fast.next == None:
            return False
        
        if slow == fast:
            return True
        return Traverse(slow.next, fast.next.next)
    return Traverse(head, head.next)


head = list.build_list([3, 2, 0, -4])
list.print_list('list : ', head)

head.next.next.next = head

print(f'has cycle : {hasCycle(head)}')